chainercv/functions/ps_roi_average_pooling_2d.py

```diff
@@ -74,47 +79,58 @@ def forward_cpu(self, inputs):
-                    out_h, stride_h, height, int(y_min))
-                if slice_h.stop <= slice_h.start:
-                    continue
-                for out_w in six.moves.range(self.out_w):
-                    slice_w, len_w = _roi_pooling_slice(
-                        out_w, stride_w, width, int(x_min))
-                    if slice_w.stop <= slice_w.start:
-                        continue
-                    for out_c in six.moves.range(self.out_c):
-                        slice_c, len_c = _roi_pooling_slice(
-                            out_c, stride_c, channels, 0)
-                        roi_data = bottom_data[
-                            batch_index, slice_c, slice_h, slice_w]\
-                            .reshape((len_c, -1))
-                        c = (out_h // group_h) * self.group_size \
-                            + (out_w // group_w)
-                        top_data[i_roi, out_c, out_h, out_w] = np.average(
-                            roi_data[c])
+        spatial_scale = self.spatial_scale
+        pooled_height = self.out_h
+        pooled_width = self.out_w
+        group_size = self.group_size
+
+        for i in six.moves.range(top_data.size):
+            n, ctop, ph, pw = np.unravel_index(i, top_data.shape)
+
+            roi_batch_ind = bottom_roi_indices[n]
+            roi_start_h = bottom_rois[n, 0] * spatial_scale
+            roi_start_w = bottom_rois[n, 1] * spatial_scale
+            roi_end_h = bottom_rois[n, 2] * spatial_scale
+            roi_end_w = bottom_rois[n, 3] * spatial_scale
+
+            roi_height = max(roi_end_h - roi_start_h, 0.1)
+            roi_width = max(roi_end_w - roi_start_w, 0.1)
+            bin_size_h = roi_height / pooled_height
+            bin_size_w = roi_width / pooled_width
+
+            hstart = int(np.floor(ph * bin_size_h + roi_start_h))
+            wstart = int(np.floor(pw * bin_size_w + roi_start_w))
+            hend = int(np.ceil((ph + 1) * bin_size_h + roi_start_h))
+            wend = int(np.ceil((pw + 1) * bin_size_w + roi_start_w))
+            hstart = min(max(hstart, 0), height)
+            wstart = min(max(wstart, 0), width)
+            hend = min(max(hend, 0), height)
+            wend = min(max(wend, 0), width)
+
+            gh = int(np.floor(ph * group_size / pooled_height))
+            gw = int(np.floor(pw * group_size / pooled_width))
+            gh = min(max(gh, 0), group_size - 1)
+            gw = min(max(gw, 0), group_size - 1)
+            c = (ctop * group_size + gh) * group_size + gw
+
+            if hstart >= hend or wstart >= wend:
+                top_data[n, ctop, ph, pw] = 0
+                continue
+
+            output_val = 0.
+            count = (hend - hstart) * (wend - wstart)
+            for y in six.moves.range(hstart, hend):
+                for x in six.moves.range(wstart, wend):
+                    output_val += bottom_data[roi_batch_ind, c, y, x]
+            output_val /= count
```

Comment:
The computation of `output_val` is no longer needed.

Reply:
updated